exonum/src/api_ng/node/public/system.rs

```diff
@@ -0,0 +1,78 @@
+// Copyright 2018 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Public system API.
+
+use api_ng::{ServiceApiScope, ServiceApiState};
+use blockchain::{Schema, SharedNodeState};
+use helpers::user_agent;
+
+#[derive(Serialize, Deserialize, PartialEq)]
+struct MemPoolInfo {
+    pub size: usize,
+}
+
+#[derive(Deserialize, Serialize, Clone, Debug, PartialEq)]
+struct HealthCheckInfo {
+    pub connectivity: bool,
+}
+
+/// Public system API.
+#[derive(Clone, Debug)]
+pub struct SystemApi {
+    shared_api_state: SharedNodeState,
+}
+
+impl SystemApi {
+    /// Creates a new `public::SystemApi` instance.
+    pub fn new(shared_api_state: SharedNodeState) -> SystemApi {
+        SystemApi { shared_api_state }
+    }
+
+    /// Adds public system API endpoints to the corresponding scope.
+    pub fn wire(self, api_scope: &mut ServiceApiScope) -> &mut ServiceApiScope {
+        self.handle_mempool_info("v1/mempool", api_scope)
+            .handle_healthcheck_info("v1/healthcheck", api_scope)
+            .handle_user_agent_info("v1/user_agent", api_scope);
+        api_scope
+    }
+
+    fn handle_mempool_info(self, name: &'static str, api_scope: &mut ServiceApiScope) -> Self {
+        api_scope.endpoint(name, move |state: &ServiceApiState, _query: ()| {
+            let snapshot = state.blockchain().snapshot();
+            let schema = Schema::new(&snapshot);
+            Ok(MemPoolInfo {
+                size: schema.transactions_pool_len(),
+            })
+        });
+        self
+    }
+
+    fn handle_healthcheck_info(self, name: &'static str, api_scope: &mut ServiceApiScope) -> Self {
+        let self_ = self.clone();
```

Comment:
Why do we need to clone `self` here?

Reply:
Because lambda captures `self`, but I want to write chain of methods and so I have to return `self` by value.

Reply:
Lambda doesn't require `self`, it requires only `self.shared_api_state.peers_info().is_empty()`.
You can simply write 
```
let connectivity = !self.shared_api_state.peers_info().is_empty();
```
and use it inside of the lambda. So you won't have to clone `self`.
In the other case I've mentioned it's also possible.